File: architecture-skills/design-patterns/examples/factory-method/factory-method.py
"""Factory Method pattern in Python.

Note: classes are first-class, so passing the transport class directly
(`plan_delivery(Truck)`) often replaces this pattern. A creator hierarchy
pays off when the creator carries real shared logic, as here.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Transport(ABC):
    """Product interface."""

    @abstractmethod
    def deliver(self, cargo: str) -> str: ...


class Truck(Transport):
    def deliver(self, cargo: str) -> str:
        return f"Truck delivers {cargo} by road in a box"


class Ship(Transport):
    def deliver(self, cargo: str) -> str:
        return f"Ship delivers {cargo} by sea in a container"


class Logistics(ABC):
    """Creator: shared workflow around one overridable creation step."""

    @abstractmethod
    def create_transport(self) -> Transport: ...

    def plan_delivery(self, cargo: str) -> str:
        transport = self.create_transport()
        return f"[{type(self).__name__}] {transport.deliver(cargo)}"


class RoadLogistics(Logistics):
    def create_transport(self) -> Transport:
        return Truck()


class SeaLogistics(Logistics):
    def create_transport(self) -> Transport:
        return Ship()


if __name__ == "__main__":
    for logistics in (RoadLogistics(), SeaLogistics()):
        print(logistics.plan_delivery("machine parts"))

File: architecture-skills/design-patterns/examples/factory-method/factory-method.js
// Factory Method pattern in JavaScript.
//
// Note: with first-class functions, passing a constructor or closure
// (`planDelivery(() => new Truck())`) often replaces the pattern. The
// class form below is worth it when the creator has real shared logic.

class Truck {
  deliver(cargo) {
    return `Truck delivers ${cargo} by road in a box`;
  }
}

class Ship {
  deliver(cargo) {
    return `Ship delivers ${cargo} by sea in a container`;
  }
}

// Creator: shared workflow around one overridable creation step.
class Logistics {
  createTransport() {
    throw new Error("Subclass must implement createTransport()");
  }

  planDelivery(cargo) {
    const transport = this.createTransport();
    return `[${this.constructor.name}] ${transport.deliver(cargo)}`;
  }
}

class RoadLogistics extends Logistics {
  createTransport() {
    return new Truck();
  }
}

class SeaLogistics extends Logistics {
  createTransport() {
    return new Ship();
  }
}

for (const logistics of [new RoadLogistics(), new SeaLogistics()]) {
  console.log(logistics.planDelivery("machine parts"));
}

// The functional alternative: the "factory method" is just a parameter.
function planDelivery(createTransport, cargo) {
  return createTransport().deliver(cargo);
}
console.log(planDelivery(() => new Ship(), "spare tires"));
